### DISCUSSION TITLE
Using MRTK without Mixed Reality Feature Tool

### DISCUSSION THREAD

**Discussion by joey-rieg-bl** [MixedRealityToolkit/MixedRealityToolkit-Unity#439] [Q&A] [answered]
Hey,
since we heavily rely on the Unity Package Manager and scoped registries for our development and CI/CD the MRFT is a big headache for us to neatly integrate MRTK in our applications.

I found the following issue opened on the old MRTK repository which has all the problems stated that our teams has as well:

https://github.com/microsoft/MixedRealityToolkit-Unity/issues/10870

Are there any news from Unity and/or a plan to utilize UPM for package installation?

**Comment by SimonDarksideJ** [comment]
I recommend updating the deployment to use UPM packages going forward, There was a reason why they were not published previously but were never formally justified.
Given I publish several other packages with similar capabilities, all using Unity dependencies like the MRTK, this should be possible with the separation from Microsoft internal.

Using the feature tool and the manual download of the packages from Azure has always been a pain, especially everytime you need to update them, given the move from MS hosted packages, this would seem reasonable.

**Comment by AMollis** [comment]
@joey-rieg-bl the recommendation for CI/CD pipelines is to add the *.TGZs via the MRFT, and then track those via your source control.  Another option is to add the packages to you own NPM feed and setup a scoped registry in your project. Internally at Microsoft, we do one of these two options.

1. Track TGZs in source control
2. Host TGZs in a private feed and use scoped registry.

Microsoft does not support using our public NPM feed outside the Mixed Reality Feature Tool (MRFT).

We are working with Unity on other methods for distributing packages.

**Reply by AMollis** [reply]
Another option, which I haven't tried, is to use UPM's Git support, https://docs.unity3d.com/Manual/upm-git.html. This won't help resolve dependencies.  If you can manually add dependencies, then this "should" work.

**Reply by joey-rieg-bl** [reply]
@AMollis thanks for the reply and the steps! :)

Those options were known to us but introduce a lot of manual handling (UPM Git dependencies, uploading newest TGZs to registry) or extra setup which are time consuming for a small team. 

We'll probably need to settle for one of those approaches but we hope that this will be much more approachable soon!

**Comment by jelmer3000** [comment]
The feature tool crashes here on start, so any alternative is highly welcome...

**Comment by keveleigh** [comment, marked as answer]
We now have a new alternative! The latest MRTK packages have been published to the Unity Asset Store: https://assetstore.unity.com/publishers/92188

These packages will show up in the Unity Package Manager and support dependency resolution, along with better visibility when updates are released. Please let us know if you run into any issues.

**Comment by dudziakl** [comment]
Dependency resolution didn't work for me, and I had to install manually all required libraries at first.  
![image](https://github.com/user-attachments/assets/4ee29924-cc3f-4fd3-a8fc-2016899d8b10)

**Reply by keveleigh** [reply]
Ohh, interesting! Thanks for the heads up. We republished a few packages to fix the version resolution, but it looks like Input may have been missed. I'll take a look and make sure this gets updated in the next release.
